File: singly_linked_list/recipe-01/c/test_string_list.c
#include "string_list.h"
#include <stdio.h>

void print1(ItemType data) {
    printf("Got this data: %s\n", data);
}

void print2(ItemType data) {
    printf("data: %s\n", data);
}

void test()
{
    printf("----------------------------%s---------------------------\n", __func__);

    SinglyLinkedList* words = NewList();
    ListAppend(words, "foo");
    ListAppend(words, "bar");
    ListAppend(words, "bim");
    ListAppend(words, "baz");
    ListAppend(words, "quux");

    printf("access by index\n");
    printf("here is a node: %s\n", ListGetItem(words, 1));
    
    printf("modify by index\n");
    ListSetItem(words, 4, "Quux");
    printf("Modified node by index: %s\n", ListGetItem(words, 4));

    printf("This list has %d elements.\n", ListCount(words));
    ListTravel(words, &print1);

    if (ListSearch(words, "foo"))
        printf("Found foo in the list.\n");
    if (ListSearch(words, "amiga"))
        printf("Found amiga in the list.\n");
    
    printf("Now we try to Delete an item\n");
    ListDelete(words, "bim");
    printf("List now has %d elements\n", ListCount(words));
    ListTravel(words, &print2);
    
    printf("Delete the first item in the list\n");
    ListDelete(words, "foo");
    printf("size: %d\n", ListCount(words));
    ListTravel(words, &print2);
    
    printf("Delete the last item in the list\n");
    ListDelete(words, "quux");
    printf("size: %d\n", ListCount(words));
    ListTravel(words, &print2);

    FreeList(words);
}

void test_reverse()
{
    printf("----------------------------%s---------------------------\n", __func__);

    SinglyLinkedList* words = NewList();
    ListAppend(words, "foo");
    ListAppend(words, "bar");
    ListAppend(words, "bim");
    ListAppend(words, "baz");
    ListAppend(words, "quux");

    printf("The origin list\n");
    ListTravel(words, &print2);

    printf("Reverse the list\n");
    ListReverse(words);
    ListTravel(words, &print2);

    FreeList(words);
}

int main()
{
    Init();

    test();
    test_reverse();

    return 0;
}

File: singly_linked_list/recipe-01/python/test_string_list.py
from singly_linked_list import SinglyLinkedList

def test():
    words = SinglyLinkedList()
    words.Append('foo')
    words.Append('bar')
    words.Append('bim')
    words.Append('baz')
    words.Append('quux')
    
    print("access by index")
    print("here is a node: {}".format(words[1]))
    
    print("modify by index")
    words[4] = "Quux"
    print("Modified node by index: {}".format(words[4]))
    
    print("This list has {} elements.".format(words.count))
    for word in words:
        print("Got this data: {}".format(word))
    
    if words.Search('foo'):
        print("Found foo in the list.")
    if words.Search('amiga'):
        print("Found amiga in the list.")
    
    print("Now we try to Delete an item")
    words.Delete('bim')
    print("List now has {} elements".format(words.count))
    for word in words:
        print("data: {}".format(word))
    
    print("Delete the first item in the list")
    words.Delete('foo')
    print("size: {}".format(words.count))
    for word in words:
        print("data: {}".format(word))
    
    print("Delete the last item in the list")
    words.Delete('quux')
    print("size: {}".format(words.count))
    for word in words:
        print("data: {}".format(word))

if __name__ == "__main__":
    test()
